Add docs instructions on how to utilize CMake Presets (#1694)
NVIDIA/cccl: CONTRIBUTING.md

## Building and Testing

> CCCL components are header-only libraries. This means there isn't a traditional build process for the library itself. However, before submitting contributions, it's a good idea to build and run tests.

Author: @mhoemmen I can fix that by linking to the dev guides https://github.com/NVIDIA/cccl/pull/1694/files#diff-eca12c0a30e25b4b46522ebf89465a03ba72a03f540796c979137931d8f92055R40.

I understand that you are asking for something more concrete maybe?

> CCCL components are header-only libraries. This means there isn't a traditional build process for the library itself. However, before submitting contributions, it's a good idea to build and run tests.

Reviewer: @gonidelis Thanks!  If you don't have an explicit policy (e.g., "new features must have tests"), then just that link would be fine.

There are multiple options for building and running our tests, which you choose depends on your preferences and if you are using [CCCL's DevContainers](.devcontainer/README.md) (highly recommended!).

> ### Using Manual Builds Scripts

Author: I have no idea how all these plural "Builds" popped up. On them.

./ci/test_cub.sh -cxx g++ -std 14 -arch "70;75;80-virtual"
```

### Using CMake Presets

> [CMake Presets](https://cmake.org/cmake/help/latest/manual/cmake-presets.7.html) are a set of configurations defined in a JSON file that specify project-wide build details for CMake. They provide a standardized and sharable way to configure, build, and test projects across different platforms and development environments. Presets are available from CMake version 3.19 and later.

Reviewer: ```suggestion
[CMake Presets](https://cmake.org/cmake/help/latest/manual/cmake-presets.7.html) are a set of configurations defined in a JSON file that specify project-wide build details for CMake. They provide a standardized and sharable way to configure, build, and test projects across different platforms and development environments. Presets are available for CMake versions 3.19 and later.
```

> There are three kinds of Presets:
> 
> - Configure Presets: specify options for the `cmake` command.
> 
> - Builds Presets: specify options for the `cmake --build` command.
> 
> - Test Presets: specify options for the `ctest` command.

Reviewer: Bullet lists have two grammatically correct options.  You can either 

1. make everything but the bullets a complete sentence, and treat the bullets as sentence fragments, or
2. treat the bullets as part of the text, and make everything a complete sentence.

In the suggestion below, I've chosen Option (2). 

```suggestion
There are three kinds of Presets.

- Configure Presets: specify options for the `cmake` command

- Build Presets: specify options for the `cmake --build` command

- Test Presets: specify options for the `ctest` command
```

> In CCCL we provide many presets to be used out of the box. You can find the complete list in our corresponding [CMakePresets.json](./CMakePresets.json) file.

Reviewer 2: There's a better way than parsing the json manually. These commands can be used to get lists of the configure, build, and test presets. They strip out the `hidden` presets that aren't intended to be used directly, and provide a customizable description of each preset.

```
cmake --list-presets # Configure presets
cmake --build --list-presets # Build presets
ctest --list-presets # Test presets
```

While there is a lot of overlap, there may be differences between the configure, build, and test presets to support various testing workflows. 

The `dev` presets are intended as a base for general development while the others are useful for screening / replicating CI failures. The non-`dev` presets are invoked by the manual build/test scripts as needed.

The CI presets change from time to time as they are tied to our CI system. They may be renamed or removed. The dev presets are more stable.

Author: Ngl, that is much needed info. Will include it. Thanks.

#### Using CMake Presets via Command Line

> Once you have created your `build` directory and navigated into it, you can configure your project for a specific preset (e.g. `thrust-cpp11`) by executing the following command:

Reviewer 2: CMake will automatically generate the preset build directories, no need to make one ahead of time.

```bash

> cmake --preset=thrust-cpp11 <path/to/cccl/root>

Reviewer 2: Rather than navigate to a build directory and pass the source to `cmake`, all of the preset commands can be executed in the root directory without passing a path:

```
git clone <cccl...>
cd cccl
cmake --preset=thrust-cpp11
cmake --build --preset=thrust-cpp11
ctest --preset=thrust-cpp11
```

That will create `build/<optional devcontainer name>/thrust-cpp11/` and build everything in there. The devcontainer name is inserted automatically on devcontainer builds to keep build artifacts separate for the different toolchains.

Reviewer 2: It's also worth mentioning that additional cmake options can still be passed in and will override the preset settings.

A common example, the presets are currently always `60;70;80` for `CMAKE_CUDA_ARCHITECTURES`, but this can be overridden at configure time with something like

```
cmake --preset=thrust-cpp20 "-DCMAKE_CUDA_ARCHITECTURES=89"
```

```

This command configures the project using the thrust-cpp11 preset. Please replace `<path/to/cccl/root>` with the actual path to the root directory of your CCCL working copy.

Upon successful configuration, initiate the build process with:

```bash

> cmake --build .

Reviewer 2: Or `cmake --build --preset=thrust-cpp11` from the source root. Either works, but will behave in slightly different ways:

`cmake --build .` from the build directory will compile every target that was configured in the configure step.

A specific build preset may only build a subset of the configured targets. Invoking `cmake --build --preset=<build_preset>` will reproduce exactly the build from CI.

The same is true for running tests. Running `ctest` in the build directory will try to run all of the tests configured in the configure step, but a preset may only run a subset of them (usually because only a subset were built). Running `ctest --preset=<test_preset>` from the source directory addresses that.

```

> 

Reviewer 3: It would be awesome to have one more paragraph showing how to run the tests. Is it just executing `ctest` or do I have to build a special test target?

Author: will add it in a subsequent commit.

#### Using CMake Presets via VS Code GUI extension (Recommended when using DevContainers)

> The recommended way to use CMake Presets is via the VS Code extension [CMake Tools](https://marketplace.visualstudio.com/items?itemName=ms-vscode.cmake-tools). As soon as you install the extension you would be able to see the sidebar menu below.

Reviewer 4: ```suggestion
The recommended way to use CMake Presets is via the VS Code extension [CMake Tools](https://marketplace.visualstudio.com/items?itemName=ms-vscode.cmake-tools) (already included in [CCCL's DevContainers](.devcontainer/README.md)). As soon as you install the extension you would be able to see the sidebar menu below.
```

   ![cmaketools sidebar](/.devcontainer/img/cmaketools_sidebar.png)

> You can specify the desire CMake Preset by clicking the "Select Configure Preset" button under the "Configure" drop down option (see image below).

Reviewer 3: ```suggestion
You can specify the desired CMake Preset by clicking the "Select Configure Preset" button under the "Configure" drop down option (see image below).
```

   ![cmaketools presets](.devcontainer/img/cmaketools_presets.png)

> After that you can select the default build target from the "Build" drop down option. As soon as you click the button a drop down menu will appear with all the available targets that are included within the preset you selected. For example if you had selected the `all-dev` preset VS Code will display all the available targets we have in cccl.

Reviewer 3: The screenshot looks like a tree widget to me, so I would use the terms node or item, instead of drop down menu.
```suggestion
After that you can select the default build target from the "Build" node. As soon as you expand it a list will appear with all the available targets that are included within the preset you selected. For example if you had selected the `all-dev` preset VS Code will display all the available targets we have in cccl.
```

   ![cmaketools presets](.devcontainer/img/cmaketools_targets.png)

You can build the selected target by pressing the gear button ![gear](.devcontainer/img/build_button.png) at the bottom of the VS Code window.

Alternatively you can select the desired target from either the "Debug" or "Launch" drop down menu (for debugging or running correspondingly). <b>In that case after you select the target and either press "Run" ![run](.devcontainer/img/run.png) or "Debug" ![debug](.devcontainer/img/debug.png) the target will build on its own before running without the user having to build it explicitly from the gear button.</b>

---

> We encourage users who want to debug device code to install the [Nsight Visual Studio Code Edition extension](https://marketplace.visualstudio.com/items?itemName=NVIDIA.nsight-vscode-edition) that enables the VS Code frontend for `cuda-gdb`. <u>To enable that you should avoid pressing the "Debug" button from the bottom menu and try launching it from the sidebar menu</u>.

Reviewer 3: ```suggestion
We encourage users who want to debug device code to install the [Nsight Visual Studio Code Edition extension](https://marketplace.visualstudio.com/items?itemName=NVIDIA.nsight-vscode-edition) that enables the VS Code frontend for `cuda-gdb`. <u>To use it you should launch from the sidebar menu instead of pressing the "Debug" button from the bottom menu.</u>.
```

Author: I like this correction.
